release-23.2: kvserver: constrain store gossip chatter

kvoli opened the pull request:
Backport 3/3 commits from #125276 on behalf of @kvoli.
Backport 1/1 commits from #125557 on behalf of @kvoli.

/cc @cockroachdb/release

----

Previously, it was possible that store's would gossip their store
descriptors rapidly on capacity changes, without limit. This patch
introduces a hard frequency limit of 2s on capacity change triggered
gossip. In other words, irrespective of any capacity changes that may
have taken place, a store will gossip at most every 2s if satisfying the
capacity delta trigger.

Resolves: #125210
Release note: None


---


Increase the relative requirement for triggering store gossip on
capacity changes for lease and ranges. Previously, the requirement was a
+-5% delta from the previously gossiped value, in addition to a minimum
(5). Increase the relative delta from 5 to 10%.

Informs: https://github.com/cockroachdb/cockroach/issues/125227
Release note: None

---

Introduce `kv.store_gossip.capacity_delta_threshold` and
`kv.store_gossip.max_frequency`, which control the capacity delta
required to trigger eager gossip for lease/range counts and the maximum
frequency of eager store gossip.

The defaults for these settings remained unchanged from the previously
hard-coded defaults:

`kv.store_gossip.capacity_delta_threshold`: 0.10
`kv.store_gossip.max_frequency`: 2s

Resolves: https://github.com/cockroachdb/cockroach/issues/125227
Release note: None

----

Release justification: Fixes unavailability caused by gossip callback congestion in many store clusters (>1k).

kvoli committed:
kvserver: limit eager store gossip frequency

Previously, it was possible that stores would gossip their store
descriptors rapidly on capacity changes, without limit. This patch
introduces a hard frequency limit of 2s on capacity change triggered
gossip. In other words, irrespective of any capacity changes that may
have taken place, a store will gossip at most every 2s if satisfying the
capacity delta trigger.

Resolves: #125210
Release note: None

kvoli committed:
kvserver: increase req for lease/range count store gossip trigger

Increase the relative requirement for triggering store gossip on
capacity changes for lease and ranges. Previously, the requirement was a
+-5% delta from the previously gossiped value, in addition to a minimum
(5). Increase the relative delta from 5 to 10%.

Informs: #125227
Release note: None

kvoli committed:
kvserver: add cluster settings for key store gossip params

Introduce `kv.store_gossip.capacity_delta_threshold` and
`kv.store_gossip.max_frequency`, which control the capacity delta
required to trigger eager gossip for lease/range counts and the maximum
frequency of eager store gossip.

The defaults for these settings remained unchanged from the previously
hard-coded defaults:

`kv.store_gossip.capacity_delta_threshold`: 0.10
`kv.store_gossip.max_frequency`: 2s

Resolves: #125227
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125425)
<!-- Reviewable:end -->

arulajmani commented:
Forgot to say, let's let this bake on master for a bit before pressing merge.

kvoli committed:
kvserver: deflake rebalance multi-store

After #125276, which reduced gossip frequency,
`TestReplicateQueueRebalanceMultiStore/multi-store` became flaky
as the queue processing speed is artificially increased but gossip
updates remain limited.

Increase the gossip frequency to deflake the test.

Fixes: #125360
Fixes: #125450
Release note: None

kvoli commented:
I'm going to include https://github.com/cockroachdb/cockroach/pull/125557 in the backport once it merges.

kvoli commented:
```
[1,469 / 1,470] Testing //pkg/sql/pgwire:pgwire_test; 41s processwrapper-sandbox
Terminated
Process exited with code 143
```

Looks unrelated, TYFTRs.
